amp/WEB-INF/src/org/digijava/kernel/validation/Validator.java

```diff
@@ -0,0 +1,221 @@
+     * @return constraint violations if any
+     */
+    public Set<ConstraintViolation> validateField(APIField field, Object value, Class<?>... groups) {
+
+        PathImpl fieldPath = advancePath(rootPath(), field);
+
+        ValidationContext validationContext = new ValidationContext();
+
+        validateField(field, value, ImmutableSet.copyOf(groups), validationContext, fieldPath);
+
+        return validationContext.getFailingConstraintViolations();
+    }
+
+    /**
+     * Validate a field.
+     *
+     * @param field field metadata
+     * @param value value of the field
+     * @param groups validation groups
+     * @param validationContext a collector for violations
+     * @param path current path
+     */
+    private void validateField(APIField field, Object value, Set<Class<?>> groups,
+            ValidationContext validationContext, PathImpl path) {
+
+        List<ConstraintDescriptor> constraintDescriptors =
+                field.getFieldConstraints().getDescriptorsFor(groups);
+
+        for (ConstraintDescriptor constraintDescriptor : constraintDescriptors) {
+            invokeConstraintValidator(field, value, validationContext, constraintDescriptor, path);
+        }
+    }
+
+    private void invokeConstraintValidator(APIField type, Object value, ValidationContext validationContext,
+            ConstraintDescriptor constraintDescriptor, PathImpl path) {
+
+        ConstraintValidator constraintValidator = instantiateAndInitialize(constraintDescriptor);
+
+        ConstraintValidatorContextImpl context = new ConstraintValidatorContextImpl(constraintDescriptor, path);
+
+        if (!constraintValidator.isValid(type, value, context)) {
+            if (context.isDefaultConstraintViolationDisabled()) {
+                validationContext.getFailingConstraintViolations().addAll(context.getCustomConstraintViolations());
+            } else {
+                ConstraintViolation violation =
+                        new ConstraintViolation(constraintValidator.getErrorMessage(), constraintDescriptor, path);
+                validationContext.getFailingConstraintViolations().add(violation);
+            }
+        }
+    }
+
+    /**
+     * Create and initialize a constraint validator.
+     *
+     * @param descriptor constraint descriptor
+     * @return constraint validator
+     */
+    private ConstraintValidator instantiateAndInitialize(ConstraintDescriptor descriptor) {
+        try {
+            ConstraintValidator validator = descriptor.getConstraintValidatorClass().newInstance();
+            validator.initialize(descriptor.getArguments());
+            return validator;
+        } catch (ReflectiveOperationException e) {
+            throw new RuntimeException("Failed to instantiate validator", e);
+        }
+    }
+
+    /**
+     * Create path for the root of the object.
+     *
+     * @return root path
+     */
+    private PathImpl rootPath() {
+        return new PathImpl();
```

Comment:
Per my understand, we should be fine to rely upon `Path` interface. If not, then may be we don't need the interface?

Reply:
PathImpl exposes some methods that are not needed in ConstriantValidator.

Reply:
We can ignore extra methods in such places. What I was wondering is the naming: we have `Path` and `PathImpl`, so having Path as an interface for abstraction layer should probably have some more purpose than ignoring methods. Though not critical, was just wondering, may be there will be purpose later.